File: backend/app/services/scoring_engine.py
from .rule_engine import analyze_with_rules, get_risk_level
from .link_analysis import analyze_urls

def compute_final_score(
    conversation: str,
    llm_result: dict | None,
    detection_mode: str,
    confidence_threshold: int,
) -> dict:
    rule_score, rule_flags = analyze_with_rules(conversation)
    link_results = analyze_urls(conversation)
    is_fallback = False
    confidence = 0.0
    llm_flags = []

    if detection_mode == "fast":
        final_score = rule_score
        explanation = "Fast mode analysis based on rule engine only."
        recommendations = ["Enable Balanced or Deep mode for LLM-powered analysis."]
        tactics = []
        category = ""
        is_fallback = True
        confidence = 0.6

    elif detection_mode in ("balanced", "deep") and llm_result:
        llm_score = llm_result.get("risk_score", rule_score)
        llm_confidence = llm_result.get("confidence", 0.8)
        llm_weight = 0.6 if detection_mode == "balanced" else 0.8
        rule_weight = 1.0 - llm_weight

        # Weighted combination
        final_score = round(rule_score * rule_weight + llm_score * llm_weight)
        category = llm_result.get("category", "")
        tactics = llm_result.get("tactics", [])
        llm_flags = llm_result.get("flagged_messages", [])
        explanation = llm_result.get("explanation", "")
        recommendations = llm_result.get("recommendations", [])

        # Override: if LLM flagged critical risk, trust it
        if llm_score >= confidence_threshold and llm_confidence >= 0.7:
            final_score = max(final_score, llm_score)
        elif llm_score >= 60 and llm_confidence >= 0.8:
            final_score = max(final_score, llm_score)

        confidence = llm_confidence

    else:
        # API key missing / LLM unavailable
        final_score = rule_score
        explanation = "Limited Analysis Mode — LLM unavailable. Results are from rule engine only."
        recommendations = ["Configure an API key in Developer Settings for full AI analysis."]
        tactics = []
        category = ""
        is_fallback = True
        confidence = 0.5

    # ─── Apply rule engine overrides (even over LLM) ──────
    # If rule engine detected direct credential harvesting, floor score
    for flag in rule_flags:
        if flag["severity"] == "critical" and "credential" in flag["reason"].lower():
            final_score = max(final_score, 80)
            if "otp" in flag["reason"].lower() or "password" in flag["reason"].lower():
                final_score = max(final_score, 82)
        elif flag["severity"] == "critical":
            final_score = max(final_score, 60)

    # ─── Blend link analysis score ──────────────────────────
    if link_results and len(link_results) > 0:
        max_link_score = max(l["risk_score"] for l in link_results)
        has_suspicious = any(l["risk_level"] != "Safe" for l in link_results)
        if has_suspicious and max_link_score > 0:
            final_score = round(final_score * 0.7 + max_link_score * 0.3)
            final_score = max(final_score, rule_score)
        if has_suspicious and "Suspicious Link Detected" not in tactics:
            tactics.append("Suspicious Link Detected")

    # Deduplicate flagged messages
    seen = set()
    all_flags = []
    for m in rule_flags + llm_flags:
        key = m.get("text", "") + m.get("reason", "")
        if key not in seen:
            seen.add(key)
            all_flags.append(m)

    risk_level = get_risk_level(final_score, confidence_threshold)

    return {
        "risk_score": final_score,
        "risk_level": risk_level,
        "category": category,
        "tactics": tactics,
        "flagged_messages": all_flags,
        "explanation": explanation,
        "recommendations": recommendations,
        "is_fallback": is_fallback,
        "confidence": round(confidence, 2),
        "link_analysis": link_results if link_results else None,
    }

File: src/services/scoringEngine.ts
import type { AnalysisResult, DeveloperSettings, FlaggedMessage, LinkAnalysis } from '../types';
import { getRiskLevel } from './ruleEngine';

export function computeFinalScore(
  ruleScore: number,
  llmResult: Partial<AnalysisResult> | null,
  settings: DeveloperSettings,
  existingFlags?: FlaggedMessage[],
  linkAnalysis?: LinkAnalysis[],
): AnalysisResult {
  let finalScore: number;
  let risk_level: AnalysisResult['risk_level'];
  let category = '';
  let tactics: string[] = [];
  let flagged_messages: AnalysisResult['flagged_messages'] = [];
  let explanation = '';
  let recommendations: string[] = [];
  let is_fallback = false;
  let confidence = 0.0;

  switch (settings.detection_mode) {
    case 'fast':
      finalScore = ruleScore;
      flagged_messages = existingFlags ?? [];
      explanation = 'Fast mode analysis based on rule engine only.';
      recommendations = ['Enable Balanced or Deep mode for LLM-powered analysis.'];
      is_fallback = true;
      confidence = 0.6;
      break;

    case 'balanced':
    case 'deep':
      if (llmResult) {
        const llmScore = llmResult.risk_score ?? ruleScore;
        const llmConfidence = llmResult.confidence ?? 0.8;
        const llmWeight = settings.detection_mode === 'balanced' ? 0.6 : 0.8;
        const ruleWeight = 1.0 - llmWeight;

        finalScore = Math.round(ruleScore * ruleWeight + llmScore * llmWeight);
        category = llmResult.category ?? '';
        tactics = llmResult.tactics ?? [];
        flagged_messages = [...new Map(
          [...(existingFlags ?? []), ...(llmResult.flagged_messages ?? [])].map(m => [m.text + m.reason, m])
        ).values()];
        explanation = llmResult.explanation ?? '';
        recommendations = llmResult.recommendations ?? [];
        confidence = llmConfidence;

        if (llmScore >= settings.confidence_threshold && llmConfidence >= 0.7) {
          finalScore = Math.max(finalScore, llmScore);
        } else if (llmScore >= 60 && llmConfidence >= 0.8) {
          finalScore = Math.max(finalScore, llmScore);
        }
      } else {
        finalScore = ruleScore;
        flagged_messages = existingFlags ?? [];
        explanation = 'Limited Analysis Mode — LLM unavailable. Results are from rule engine only.';
        recommendations = ['Configure an API key in Developer Settings for full AI analysis.'];
        is_fallback = true;
        confidence = 0.5;
      }
      break;

    default:
      finalScore = ruleScore;
      flagged_messages = existingFlags ?? [];
      explanation = 'Limited Analysis Mode — unknown detection mode.';
      recommendations = [];
      is_fallback = true;
      confidence = 0.5;
  }

  if (linkAnalysis && linkAnalysis.length > 0) {
    const maxLinkScore = Math.max(...linkAnalysis.map(l => l.risk_score));
    const hasSuspicious = linkAnalysis.some(l => l.risk_level !== 'Safe');
    if (hasSuspicious && maxLinkScore > 0) {
      finalScore = Math.round(finalScore * 0.7 + maxLinkScore * 0.3);
      finalScore = Math.max(finalScore, ruleScore);
    }
    if (!tactics.includes('Suspicious Link Detected') && hasSuspicious) {
      tactics.push('Suspicious Link Detected');
    }
  }

  const threshold = settings.confidence_threshold;
  risk_level = getRiskLevel(finalScore, threshold);

  return {
    risk_score: finalScore,
    risk_level,
    category,
    tactics,
    flagged_messages,
    explanation,
    recommendations,
    is_fallback,
    confidence,
    link_analysis: linkAnalysis && linkAnalysis.length > 0 ? linkAnalysis : undefined,
  };
}
